sw/device/lib/testing/dif_uart_test.cc

```diff
@@ -0,0 +1,819 @@
+}
+
+// dif_uart_irqs_disable() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, IrqsDisableFailureUartNull) {
+  uint32_t state;
+  EXPECT_FALSE(dif_uart_irqs_disable(nullptr, &state));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, IrqsDisableSuccess) {
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET,
+                 0b00000000'00000000'00000000'00000000);
+
+  EXPECT_TRUE(dif_uart_irqs_disable(&dif_uart_, nullptr));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, IrqsDisableStateSuccess) {
+  // IRQs disabled
+  EXPECT_READ32(UART_INTR_ENABLE_REG_OFFSET,
+                0b00000000'00000000'00000000'00000000);
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET,
+                 0b00000000'00000000'00000000'00000000);
+
+  uint32_t state = 0b11111111'11111111'11111111'11111111;
+  EXPECT_TRUE(dif_uart_irqs_disable(&dif_uart_, &state));
+
+  EXPECT_EQ(state, 0b00000000'00000000'00000000'00000000);
+
+  // IRQs enabled
+  EXPECT_READ32(UART_INTR_ENABLE_REG_OFFSET,
+                0b11111111'11111111'11111111'11111111);
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET,
+                 0b00000000'00000000'00000000'00000000);
+
+  state = 0b00000000'00000000'00000000'00000000;
+  EXPECT_TRUE(dif_uart_irqs_disable(&dif_uart_, &state));
+
+  EXPECT_EQ(state, 0b11111111'11111111'11111111'11111111);
+}
+
+// dif_uart_irqs_restore() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, IrqsRestoreFailureUartNull) {
+  EXPECT_FALSE(dif_uart_irqs_restore(nullptr, 0xffffffff));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInitialised, IrqsRestoreSuccess) {
+  uint32_t state = 0b11111111'11111111'11111111'11111111;
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, state);
+
+  EXPECT_TRUE(dif_uart_irqs_restore(&dif_uart_, state));
+
+  state = 0b00000000'00000000'00000000'00000000;
+  EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, state);
+
+  EXPECT_TRUE(dif_uart_irqs_restore(&dif_uart_, state));
+}
+
+// dif_uart_irq_enable() tests
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInterrupt, IrqEnableFailureUartNull) {
+  EXPECT_FALSE(dif_uart_irq_enable(nullptr, kDifUartInterruptTxWatermark,
+                                   kDifUartEnable));
+}
+
+/**
+ * TODO
+ */
+TEST_F(DifUartTestUartInterrupt, IrqEnableSuccess) {
+  // Enable interrupts
+  for (auto const& irq : uart_irq_reg_map_) {
+    EXPECT_READ32(UART_INTR_ENABLE_REG_OFFSET,
+                  0b00000000'00000000'00000000'00000000);
+    EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, irq.second);
+
+    EXPECT_TRUE(dif_uart_irq_enable(&dif_uart_, irq.first, kDifUartEnable));
+  }
+
+  // Disable interrupts
+  for (auto const& irq : uart_irq_reg_map_) {
+    EXPECT_READ32(UART_INTR_ENABLE_REG_OFFSET,
+                  0b11111111'11111111'11111111'11111111);
+    EXPECT_WRITE32(UART_INTR_ENABLE_REG_OFFSET, ~irq.second);
+
+    EXPECT_TRUE(dif_uart_irq_enable(&dif_uart_, irq.first, kDifUartDisable));
+  }
+
+  EXPECT_EQ(uart_irq_reg_map_.size(), 8); // TODO
```

Comment:
What does this assertion actually ensure?

Reply:
This has been deleted. This expectation now lives in `CheckAllInterruptEnumValuesAreCovered`. It just checks whether all of the enum variants are covered - essentially compares `dif_uart.h` num of enum variants with the hardcoded value of how many enum variants are there at the time of writing this test. This way if value is added, it will result in mismatch, and we will know that it hasn't been tested.